Выполнить ТЗ спринта №8

SergeyLeshchev opened the pull request:
1. Добавить новые поля в классы задач
2. Создать методы, которые рассчитывают startTime и duration у Epic
3. Обновить функцию сохранения данных в файл
4. Добавить функцию учета приоритета задач
5. Оптимизировать сортировку задач по приоритету
6. Реализовать проверку пересечения startTime задач
7. Заменить циклы на потоки
8. Внести изменения в существующие методы для правильной работы программы
9. Исправить и добавить тесты

Commit:
Выполнить ТЗ спринта №8

1. Добавить новые поля в классы задач
2. Создать методы, которые рассчитывают startTime и duration у Epic
3. Обновить функцию сохранения данных в файл
4. Добавить функцию учета приоритета задач
5. Оптимизировать сортировку задач по приоритету
6. Реализовать проверку пересечения startTime задач
7. Заменить циклы на потоки
8. Внести изменения в существующие методы для правильной работы программы
9. Исправить и добавить тесты

Commit:
Выполнить ТЗ спринта №8

1. Внести исправления код-стайл

Commit:
Выполнить ТЗ спринта №8

1. Внести исправления №2 код-стайл

Ivanov-s-tmn commented on src/data/Epic.java:
```diff
@@ -1,29 +1,45 @@
+    @Override
+    public LocalDateTime getEndTime() {
+        return endTime;
+        //return (endTime != null) ? endTime : getStartTime();
```
Лишнее удаляй

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -161,42 +219,60 @@ public Subtask getSubtask(int idOfTask) {
         return historyManager.getHistory();
     }
+
+    @Override
+    public TreeSet<Task> getPrioritizedTasks() {
+        return prioritizedTasks;
+    }
```
Возвращай списком, он сохраняет порядок отсортированного списка. Где возвращается не один объект, обычно всегда используют список, с ним проще работать

SergeyLeshchev replied:
Согласен, в тестовом методе у меня в итоге криво получилось, потому что пытаюсь сравнить дерево и список. Теперь можно исправить, чтобы было красиво и удобно))

SergeyLeshchev replied:
И в некоторых других методах идея ругалась, что дерево не итерабл, поток не может реализовать некоторые функции, или я не могу. Список однозначно удобнее))

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -161,42 +219,60 @@ public Subtask getSubtask(int idOfTask) {
+
+    private boolean isTaskCross(Task task1, Task task2) {
+        // Либо старт первой задачи находится между стартом и финишем второй задачи
+        return (task1.getStartTime().isAfter(task2.getStartTime()) && task1.getStartTime().isBefore(task2.getEndTime())) ||
+                // Либо финиш первой задачи находится между стартом и финишем второй задачи
+                (task1.getEndTime().isAfter(task2.getStartTime()) && task1.getEndTime().isBefore(task2.getEndTime()));
```
Это условие не покрывает все случаи, не обязательно же старт или конец одной задачи должны находиться в интервале между временем старта и конца существующей задачи. Интервал может перекрываться и полностью (время старта 1 задачи до времени старта 2 и время окончания 1 после времени окончания 2)

Самое простое - это написать условие для не пересекающихся интервалов.
     Непересечение, это когда начало одного интервала больше конца другого интервала, или конец одного интервала меньше начала другого:
              a.start > b.end OR a.end < b.start
    Соответственно, пересечение - это отрицание того, что мы получили - т.е. если для интервалов выполняется:
             NOT( a.start > b.end OR a.end < b.start )
    Теперь раскрываем по правилам логики скобки и получаем a.start <= b.end AND a.end >= b.start
останется превратить это псевдоусловие в логическое условие на языке java

SergeyLeshchev replied:
спасибо за такое развернутое объяснение. Понял в чем проблема, но не очень уверенно чувствую себя в этой теме. Изучил еще информацию, но чувствую, что плаваю, надеюсь правильно сделал)

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -111,31 +166,34 @@ public ArrayList<Subtask> getAllSubtasks() {
         }
+
         subtaskCollection.clear();
         epicCollection.clear();
```
Нужно не забыть удалить подзадачи из истории и отсортированного списка при удалении эпиков

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -63,7 +89,17 @@ public void addSubtask(Subtask subtask) {
+            if (isValidTime(subtask) && prioritizedTasks.stream()
+                    .noneMatch(subtask1 -> isTaskCross(subtask, subtask1))) {
+                // удаляем старый экземпляр задачи, если в списке есть задача с этим же id
+                if (prioritizedTasks.contains(subtask)) {
+                    prioritizedTasks.remove(subtask);
+                }
```
На будущее, у нас `equals()` реализован только по айди и такой вариант реализации удаления подойдет, но если бы был equals() по всем полям, то такой вариант бы не отработал, так как в метод приходит уже измененный объект `subtask` и нужно удалять объект до изменения (доставать из хешмапы до добавления)

SergeyLeshchev replied:
Да, тоже думал об этом, спасибо, что обратил внимание и предложил хорошее решение, запомню эту крутую механику :)

Ivanov-s-tmn commented on test/manager/InMemoryTaskManagerTest.java:
Смотри. для тестов таск менеджера и файлового менеджера нужно реализовать абстрактный базовый класс, в котором будет реализация всех тестов для методов интерфейса `TaskManager`

Благодаря параметризации `<T extends TaskManager>` мы задаем для наследников `TaskManagerTest` ограничения по типам, которые может принимать параметризованное поле (т.е. это наследники `TaskManager`)
```java
public abstract class TaskManagerTest<T extends TaskManager> {
     protected T taskManager; //параметризованное поле taskManager, которое мы можем использовать в наследниках и инициализировать его конкретной реализацией таск менеджера

     /**
     * Реализация тестов всех методов интерфейса `TaskManager` должна быть в этом классе, классы-наследники уже будут их выполнять с конкретной реализацией менеджера
     */

}
```
Как пример:
```java
public class InMemoryTaskManagerTest extends TaskManagerTest<InMemoryTaskManager> {

    @BeforeEach
    void beforeEach() {
        taskManager = new InMemoryTaskManager(); //инициализируем T taskManager реализацией InMemoryTaskManager
        initTasks(); //Можно вынести создание тестовых задач в TaskManagerTest
    }
}
```

С тестами файлового менеджера ситуация аналогичная, только инициализировать таск менеджер будет реализацией `FileBackedTaskManager` и будут дополнительные тесты его функционала (проверка восстановления состояния из файла). `InMemoryTaskManagerTest` по идее может быть пустым и прогонять тесты `TaskManagerTest`, потому что в нем нет особого функционала как у файлового менеджера

SergeyLeshchev replied:
Мне пришлось убрать генерацию задач в тестах в BeforeEach, потому что в абстрактном классе мне кажется не нужны поля в виде задач для всех наследников. А если создавать задачи внутри метода BeforeEach, то другие методы их не видят. В InMemoryTaskManagerTest изначально были задачи, но потом я увидел, что тесты используют только один эпик и лишь в одном или двух местах, подумал, что в таком случае это тоже становится не актуальным. Для FileBackedTaskManager это тоже не нужно

Ivanov-s-tmn replied:
В целом, можно и так, но мне кажется очень много дублирования, которого можно избежать в тестах и сделать их более компактными и понятными, создание задач и добавление в таск менеджер как раз хорошо подходит для этого:
```java
Task task = new Task("task", "task description", Status.NEW, LocalDateTime.parse("2003-12-21T21:21:21"), Duration.ofMinutes(10));
taskManager.addTask(task);
```
А в тестах можно использовать одни и те же объекты по сути, нет необходимости создавать какие-то специфичные (только в крайних случаях)

Commit:
Выполнить работу над ошибками в ТЗ спринта №8

1. Изменить getPrioritizedTasks(), метод возвращает ArrayList
2. Исправить неправильную логику метода isTaskCross()
3. Исправить метод removeAllEpics()
4. Добавить тестовый класс TaskManagerTest
5. Изменить тестирование согласно ТЗ
